tools/xdp/s2n-quic-xdp/src/syscall.rs

```diff
@@ -0,0 +1,342 @@
+    }
+
+    // the cookie syscall is not supported here so return the default value of 1
+    // https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L55
+    if errno::errno().0 == libc::ENOPROTOOPT {
+        return Ok(1);
+    }
+
+    Err(io::Error::last_os_error())
+}
+
+/// Configures the fill ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L257).
+#[inline]
+pub fn set_fill_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemFillRing, &len)
+}
+
+/// Configures the completion ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L263).
+#[inline]
+pub fn set_completion_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemCompletionRing, &len)
+}
+
+/// Configures the RX ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1081).
+#[inline]
+pub fn set_rx_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::RxRing, &len)
+}
+
+/// Configures the TX ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1093).
+#[inline]
+pub fn set_tx_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::TxRing, &len)
+}
+
+/// Configures the UMEM options for the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L344).
+#[inline]
+pub fn set_umem<Fd: AsRawFd>(fd: &Fd, reg: &UmemReg) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemReg, reg)
+}
+
+/// Binds the provided AF_XDP socket to an address
+#[inline]
+pub fn bind<Fd: AsRawFd>(fd: &Fd, addr: &mut Address) -> Result<()> {
+    libc!(bind(
+        fd.as_raw_fd(),
+        addr as *mut _ as _,
+        size_of::<Address>() as _
+    ))?;
+    Ok(())
+}
+
+/// Notifies the kernel to send any packets on the TX ring
+///
+/// This should be called after checking if the TX ring needs a wake up.
+#[inline]
+pub fn wake_tx<Fd: AsRawFd>(fd: &Fd) -> Result<()> {
+    libc!(sendto(
+        fd.as_raw_fd(),
+        core::ptr::null_mut(),
+        0,
+        libc::MSG_DONTWAIT,
+        core::ptr::null_mut(),
+        0,
+    ))?;
+    Ok(())
+}
+
+/// Tries to receive packets on the RX ring
+///
+/// Note that this call is non-blocking and is usually used with busy polling. Use [`libc::poll`]
+/// instead to block a task on socket readiness.
+#[inline]
+pub fn busy_poll<Fd: AsRawFd>(fd: &Fd) -> Result<u32> {
```

Comment:
Wondering if this could return `Poll` and maybe should be called poll_rx. And maybe similar for above?

Reply:
Hmm I guess we could return `Poll`. It would be the same thing as just checking for `0` so I'm not sure it's worth it. Essentially the receiver will be doing this in a hot loop: 

```rust
loop {
    while syscall::busy_poll(&fd)? == 0 {}

    // process new entries
}
```